### DISCUSSION TITLE
using mqtt events with zones for direction of motion / selecting images

### DISCUSSION THREAD

**Discussion by notroblowe** [blakeblackshear/frigate#1839] [Ask A Question]
I'm currently using 0.9rc4 with HA. 

I've been struggling a bit to come up with reliable automations based on direction of motion. My automations based on presence in a zone are working great. 

I've done some searching in the forum but haven't really found a good description of the mqtt event message format. 

Each event contains a before and an after section. 

Within those sections are current and entered zones. 

I have zones defined on a camera, say street, driveway, walkway and step. I want to get a notification when a car enters the driveway from the street, or when a person goes to the step from the walkway. 

Looking at some examples, I thought that I should check the before/current_zones for street, and then the after/current_zones for driveway if I want to have an automation when a car enter the driveway, but I can't really seem to track a pattern on how/when current or entered zones are being updated. 

What is the expected behaviour in a 

Is it possible to log the generated mqtt events in frigate so that I can better analyze the data that is being generated? I currently use MQTT Explorer to subscribe to the event topic, but that doesn't allow for saving of the events. 

Also, I've been using http://ccab4aaf-frigate-beta:5000/api/events/{{trigger.payload_json["after"]["id"]}}/thumbnail.jpg to add an image to the telegram messages that I send in the automations. I often get two or three copies of the same image from update events with this approach. Is there something I can trigger off of in the mqtt events to know if the image will be a duplicate, or a better approach to handling images in these notifications? I don't mind the multiple notifications when a detection is in progress, but I'd like to see more relevant images in them if they are available. 

This has been a lot of fun to play with an I'm hoping to learn more as I experiment with it more.

**Comment by tve** [comment]
Duunno if this helps, but here's a code snippet from my processing of events in node-red (so it's javascript). `pl` is the event and I have an "inside" zone and an "outside" zone, and I want "entering" and "leaving" messages, plus "at entrance" for stuff that moves on the inside but doesn't get outside:
```
// what happened
let inside_entered = pl.after.entered_zones.includes("inside")
let inside_now = pl.after.current_zones.includes("inside")
let outside_entered = pl.after.entered_zones.includes("outside")
let outside_now = pl.after.current_zones.includes("outside")

// if they're only outside, don't care
if (!inside_entered) return null
// skip clip-ready events
if (pl.type == "clip_ready") return null
// determine "action"
let what = outside_entered && inside_now ? "entering"
         : outside_now && inside_entered ? "leaving"
         : "at entrance"
// if they're hanging around at the entrance delay 'til end event: perhaps
// we find out they're entering or leaving by then
if (what == "at entrance" && !pl.after.end_time) return null
```
I do keep track of event IDs for which I've sent an alert so I don't send multiple.
